examples/agent_policy.py

```diff
@@ -437,17 +443,44 @@ def action_code_to_action(self, action_code, game, unit=None, city_tile=None, te
             elif unit is not None:
                 x = unit.pos.x
                 y = unit.pos.y
-
-            return self.actionSpaceMap[action_code](
-                game=game,
-                unit_id=unit.id if unit else None,
-                unit=unit,
-                city_id=city_tile.city_id if city_tile else None,
-                citytile=city_tile,
-                team=team,
-                x=x,
-                y=y
-            )
+            
+            if city_tile != None:
+                action =  self.actions_cities[action_code%len(self.actions_cities)](
+                    game=game,
+                    unit_id=unit.id if unit else None,
+                    unit=unit,
+                    city_id=city_tile.city_id if city_tile else None,
+                    citytile=city_tile,
+                    team=team,
+                    x=x,
+                    y=y
+                )
+
+                # If the city action is invalid, default to research action automatically
+                if not action.is_valid(game, actions_validated=[]):
```

Comment:
Should we have a `do_nothing` action for the city? There can be cases where the city should wait before spawning a unit and keep its cooldown at 0. What do you think?

Reply:
Good suggestion. A smart AI should be able to learn to use it some rare cases.

Reply:
Sorry, I'm just going to leave this one for future work so I don't have to retest everything. I think someone adding this to their personal scripts is a good idea, or you are welcome to add it to the example if you like!